File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.java
class Solution {
    public int maxProfit(int[] prices) {
        int l = 0;
        int maxProfit = 0;

        for(int r = 0; r < prices.length; r++) {
            if(prices[r] > prices[l]) {
                int profit = prices[r] - prices[l];
                maxProfit = Math.max(maxProfit, profit);
            } else {
                l = r;
            }
        }
        return maxProfit;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-4.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def checkEatingSpeed(k: int) -> int:
            hours = 0
            for p in piles:
                hours += math.ceil(p / k)

            return hours

        # l pointer = 1, can't eat 0 banana
        # r pointer = max(piles), fastest way to eat each pile
        l, r = 1, max(piles)

        while l < r:
            k = l + (r - l) // 2
            hours = checkEatingSpeed(k)

            if hours > h:
                l = k + 1
            else:
                r = k
        
        return l